### ISSUE TITLE
Post Combat Stats

### ISSUE THREAD

**Issue by coreyjkelly** [cynicaloptimist/improved-initiative#423]
An enhancement request. In the interest of motivating my players to keep their turns efficient, it would be nice to have some simple stats about how much time a player takes per turn, on average, that are displayed at the end of combat. Since we've already got a round timer, this requires relatively little work to implement.

I've got a basic working example ready for a PR if you think there would be interest in this.

**Comment by cynicaloptimist** [comment]
Yup, I'd take a look at the PR if some test coverage for the new feature is included. What's the UI for this look like?

**Comment by coreyjkelly** [comment]
Once the PlayerView implementation is done, I'll start thinking about what would be required for adding damage tracking to this, since most of the infrastructure will be the same as the time tracking.

Obviously only useful/accurate for groups who exclusively use suggested damage. Otherwise, could have an optional "detailed tracking" mode with a Prompt to select which Combatant dealt the damage being applied?

**Comment by coreyjkelly** [comment]
So, tracking/reporting damage and healing done _to_ the players is relatively easy. Damage/healing done _by_ them is a bit harder, and is unfortunately the thing that more players will want to know. What do you think of this:

- When the "Post Combat Stats" setting is on, a dropdown is added to `ApplyDamagePrompt`, asking which `Combatant` dealt the damage. This would default to the active `Combatant`, and would only include player characters.

The one hitch I see: I know that I often apply damage/healing to the wrong combatant. I "undo" this by doing the opposite. If we're tracking all of this, those would be separately tracked as damage **and** healing. Fixing this problem would require some Undo functionality, at least for damage/healing.

**Comment by cynicaloptimist** [comment]
I've thought about this a bit and I agree with your analysis. If users want accurate reporting, they'll need to manually ensure the right 'attacker' is selected for any applied damage, which I think is going to slow people down and make the app experience a bit worse. I'd prefer to hold off on this unless you think you can make some really compelling UI for it.

**Comment by cynicaloptimist** [comment]
As far as Undo functionality, I have long dreamt of moving the entire EncounterState into a Redux store, and have all in-encounter actions defined as redux actions. This would be an excellent way to seperate the encounter domain model from the UI models, and also makes it really easy to implement an undo/redo stack.

**Comment by coreyjkelly** [comment]
I agree, but I'll poke around and see how the UI feels. I don't mind putting something together and seeing if we can make it work.

I think for groups who want the detailed stats, it might be worth the extra bit of effort. As long as those prompts are only displayed for users who have enabled that setting. The default values in the resultant prompts will be very important. It's relatively rare that damage/healing is done by a combatant other than the active one (I can only think of Opportunity Attacks, Legendary Actions, and held Actions in 5e). As long as active combatant is the default, it usually won't require more than the one click to dismiss the prompt.

I ran a session last weekend and was super excited about the time stats. Read them out to players at the end of a ~3 hour combat and their immediate response was "Can you tell us how much damage we did?!"... So I'm a bit biased in my motives here :)

**Comment by coreyjkelly** [comment]
The way I'd envisioned this turned out to be a bit more complex than I was able to handle. Might be an easy problem for someone else if there's interest in the future!

The original intent of this issue was time stats, so I'll close it now!

**Comment by cynicaloptimist** [comment]
Sounds good, thanks very much for cleaning this up!
